src/document_parser_callbacks.h

```diff
@@ -89,20 +89,45 @@ really_inline bool parser::on_null_atom() noexcept {
 really_inline uint8_t *parser::on_start_string() noexcept {
   /* we advance the point, accounting for the fact that we have a NULL
     * termination         */
+  // If we limit JSON documents to strictly less 4GB of
+  // string content, then current_string_buf_loc
+  // - doc.string_buf.get() fits in 32 bits. This leaves us
+  // three free bytes.
   write_tape(current_string_buf_loc - doc.string_buf.get(), internal::tape_type::STRING);
-  return current_string_buf_loc + sizeof(uint32_t);
+  return current_string_buf_loc + sizeof(uint16_t);
 }
 
 really_inline bool parser::on_end_string(uint8_t *dst) noexcept {
   uint32_t str_length = uint32_t(dst - (current_string_buf_loc + sizeof(uint32_t)));
   // TODO check for overflow in case someone has a crazy string (>=4GB?)
   // But only add the overflow check when the document itself exceeds 4GB
   // Currently unneeded because we refuse to parse docs larger or equal to 4GB.
-  memcpy(current_string_buf_loc, &str_length, sizeof(uint32_t));
+
+  // We have two scenarios here. Either the string length is
+  // less than 0x7fffff in which case, we have room in the string
+  // header and all is good. Otherwise, we can encode the
+  // string length in the document itself, taking care to
+  // ensure that we do so in ASCII.
+  if(likely(str_length <= 0x7fffff)) { // likely
+    doc.tape[current_loc-1] |=  uint64_t(str_length) << 32;
+    // we have a string header that must be ASCII, unused in
+    // this common case
+    current_string_buf_loc[0] = uint8_t(32); // space
```

Comment:
AFAIK, the actual bytes 0-31 are valid UTF-8 as well. We don't check for them, at any rate :) If it makes this simpler at all.

Reply:
They are. My thinking was "what if someone reads this in a debugger or something"...

Reply:
OK. Added null characters.